File: frontend/src/models/Pattern.model.ts
import { Dayjs } from 'dayjs';

import type { ICreator } from '@/models/Creator.model';
import type { IUser } from '@/models/User.model';

export interface IPattern {
    readonly reference: string;
    readonly createdAt: Dayjs;
    title: string;
    readonly width: number;
    readonly height: number;
    readonly price: number;
    readonly thumbnailUrl: string;
    readonly threadCount: number;
    readonly stitchCount: number;
    readonly bannerImageUrl: string | null;
    externalShopUrl: string | null;
    readonly titleSlug: string;
    aidaCount: number;
    readonly user: IUser;
    readonly creator: ICreator | null;
}

export interface IStitch {
    readonly threadIndex: number;
    readonly x: number;
    readonly y: number;
    stitchedAt: Dayjs | null;
}

export interface IPatternThread {
    readonly index: number;
    readonly name: string;
    readonly description: string;
    readonly colour: string;
}

File: backend/ClickStitch/Models/PatternModel.cs
namespace ClickStitch.Models;

public sealed class PatternModel
{
    public required Guid Reference { get; init; }
    public required DateTime CreatedAt { get; init; }
    public required string Title { get; init; }
    public required int Width { get; init; }
    public required int Height { get; init; }
    public required decimal Price { get; init; }
    public required string? ThumbnailUrl { get; init; }
    public required int ThreadCount { get; init; }
    public required int StitchCount { get; init; }
    public required string BannerImageUrl { get; init; }
    public required string? ExternalShopUrl { get; init; }
    public required string TitleSlug { get; init; }
    public required int AidaCount { get; init; }
    public required CreatorModel? Creator { get; init; }
    public required UserModel User { get; init; }
}

public sealed class PatternThreadModel
{
    public required int Index { get; init; }
    public required string Name { get; init; }
    public required string Description { get; init; }
    public required string Colour { get; init; }
}